enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,613 @@
+  healthy? -- Yes --> do-persist-version
+  healthy? -. No  .-> red
+
+  exit -.-> red
+  red --> reboot
+```
+
+### API Extensions
+
+None
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+Since this is a must-have functionality, the risks are not foreseeing fail scenarios in advance and implementation bugs.
+
+To mitigate the risks, a thorough review of the enhancement must be done by MicroShift, OpenShift, and RHEL teams,
+and making sure testing strategy is sound and prioritized equally with the feature development.
+
+### Drawbacks
+
+N/A
+
+## Design Details
+
+### Definitions
+
+- **ostree commit**: TODO
+- **ostree deployment**: TODO
+- **Rollback**: booting previous ostree commit due to greenboot assessing system to not be functional
+- **Upgrade**: running newer version of MicroShift than previously as a result of booting another ostree commit
+- **Downgrade**: running older version of MicroShift than previously as a result of booting another ostree commit
+- **Backup**: backing up `/var/lib/microshift`
+- **Restore**: restoring `/var/lib/microshift`
+- **Version Metadata**: File residing in MicroShift data dir containing version of MicroShift that successfully started (cluster was healthy)
+- **MicroShift greenboot healthcheck**: Program verifying the status of MicroShift's cluster
+
+### Preface
+
+Every action related to procedure described in this enhancement is 
+performed after system's boot rather than immediately before shutdown. 
+Greenboot's healthchecks, green and red scripts are executed independent of MicroShift's processes.
+Actions related to backup, restore, and data migration are performed with MicroShift 
+partially running, i.e. only etcd and kube-apiserver are running.
+
+Only one backup of MicroShift data will be stored at a given moment
+due to high probability of devices having limited storage.
+
+### System rollback and failed data restore detection
+
+TODO: How to not fall into endless loop of rollbacks 
+
+### Integration with greenboot
+
+Depending on result of greenboot's healthcheck either "green" (successful boot) or "red" (unsuccessful) scripts are executed before rebooting the system.
+
+MicroShift will integrate with that system to persist an action to perform on next boot:
+- "green": on next boot, before MicroShift starts, make a backup of MicroShift's data
+- "red": on next boot, before MicroShift starts, restore MicroShift's data from backup
+
+Functionality will be implemented by placing in `/etc/greenboot/green.d` and `/etc/greenboot/red.d`
+bash scripts containing with simple logic or, if needed,
+executing commands `microshift greenboot green` and `microshift greenboot red` in case of needing to put more information into file with action for next boot.
+Alternatively, a symlinks to `microshift` binary can be made and MicroShift modified to run specific command depending on content of argv[0] (just like BusyBox).
```

Comment:
I think this paragraph is considering 2 locations for the implementation, either in the main microshift binary or a separate shell script just for integrating with the greenboot triggers. Do you see a significant benefit to choosing either approach?

Reply:
Bash scripts will be simpler, it could be just `echo restore/backup > /var/lib/microshift.next-boot-action`

However, I feel like providing command in `microshift` will allow us to actually test that it does what we want (virtual filesystem + unit tests) and keep all paths in sync across green, red, and backup/restore functions because they would reference the same string object in the binary.
Same thing for the contents of the file, mentioned functions will be able to use strictly defined strings (as enum)  to serialization/deserialization is in sync